packages/material-ui/src/Menu/Menu.js

```diff
@@ -115,37 +168,119 @@ const Menu = React.forwardRef(function Menu(props, ref) {
     }
   });
 
+  const handleSetOpenSubMenuIndex = (value) => {
+    if (value === null) {
+      if (contentAnchorRef.current.parentElement) {
+        contentAnchorRef.current.parentElement.children[openSubMenuIndex].focus();
+      }
+    }
+    setOpenSubMenuIndex(value);
+  };
+
   const items = React.Children.map(children, (child, index) => {
+    if (!React.isValidElement(child)) {
+      return;
+    }
+
+    const { subMenu, onMouseMove: onMouseMoveChildProp } = child.props;
+    const { anchorEl } = other;
+
+    const hasSubMenu = Boolean(subMenu);
+    const parentMenuOpen = Boolean(anchorEl);
+
+    let additionalPropsAdded = false;
+    const additionalProps = {};
+
+    // This is the original purpose of this React.Children.map and is basically unchanged.
     if (index === activeItemIndex) {
-      return React.cloneElement(child, {
+      additionalPropsAdded = true;
+
+      Object.assign(additionalProps, {
         ref: (instance) => {
           contentAnchorRef.current = instance;
           setRef(child.ref, instance);
         },
       });
     }
 
+    // If the current Menu item in this map has a subMenu,
+    // we need the parent Menu to orchestrate its subMenu
+    if (hasSubMenu && parentMenuOpen) {
+      additionalPropsAdded = true;
+
+      const handleArrowRightKeydown = (event) => {
+        if (event.key === 'ArrowRight') {
+          event.preventDefault();
+          setOpenSubMenuIndex(index);
+        }
+      };
+
+      Object.assign(additionalProps, {
+        handleArrowRightKeydown,
+        openSubMenu: index === openSubMenuIndex && !entering,
+        setParentOpenSubMenuIndex: handleSetOpenSubMenuIndex,
+      });
+    }
+
+    // If there are ANY children with subMenus, then ALL
+    // of the children need to know how to close any open subMenus
+    // and reset the state that controls which subMenu is open.
+    if (atLeastOneSubMenu) {
+      additionalPropsAdded = true;
+
+      Object.assign(additionalProps, {
+        onMouseMove: (e) => {
+          setOpenSubMenuIndex(index);
+          if (onMouseMoveChildProp) {
+            onMouseMoveChildProp(e);
+          }
+        },
+        handleParentClose: handleOnClose,
+      });
+    }
+
+    // Using a semaphore instead of inspecting addtionalProps
+    // directly to avoid performance hits at scale. Might be
```

Comment:
There's no need to address performance prematurely. First we make it work, then we measure, then we make it fast.

Reply:
Fair. Is there an action you'd like me to take based on this feedback, or is the semaphore still fine with you. It does "work", but maybe you think it adds unnecessary noise in the code? Or is the semaphore pattern fine?

Reply:
Remove the semaphore I guess. We never relied on this pattern before and I suspect we'll never will. Otherwise this code will just look like it shouldn't be touched when we actually never knew why we added it.

Reply:
Cool, I'll circle back on this once I get a green CI.

Reply:
Refactored this and pushed.